scripts/bumper/main.go

```diff
@@ -0,0 +1,320 @@
+	}
+
+	logLevel, _ := logrus.ParseLevel(opts.logLevel)
+	logger.SetLevel(logLevel)
+
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
+	defer cancel()
+
+	var commits []commit
+	var err error
+	if opts.commitFileInput != "" {
+		rawCommits, err := os.ReadFile(opts.commitFileInput)
+		if err != nil {
+			logrus.WithError(err).Fatal("could not read input file")
+		}
+		if err := json.Unmarshal(rawCommits, &commits); err != nil {
+			logrus.WithError(err).Fatal("could not unmarshal input commits")
+		}
+	} else {
+		commits, err = detectNewCommits(ctx, logger.WithField("phase", "detect"), opts.stagingDir)
+		if err != nil {
+			logger.WithError(err).Fatal("failed to detect commits")
+		}
+	}
+
+	if opts.commitFileOutput != "" {
+		commitsJson, err := json.Marshal(commits)
+		if err != nil {
+			logrus.WithError(err).Fatal("could not marshal commits")
+		}
+		if err := os.WriteFile(opts.commitFileOutput, commitsJson, 0666); err != nil {
+			logrus.WithError(err).Fatal("could not write commits")
+		}
+	}
+
+	switch mode(opts.mode) {
+	case summarize:
+		writer := tabwriter.NewWriter(os.Stdout, 0, 4, 2, ' ', 0)
+		for _, commit := range commits {
+			if _, err := fmt.Fprintln(writer, commit.Date.Format(time.DateTime)+"\t"+"operator-framework/"+commit.Repo+"\t", commit.Hash+"\t"+commit.Author+"\t"+commit.Message); err != nil {
+				logger.WithError(err).Error("failed to write output")
+			}
+		}
+		if err := writer.Flush(); err != nil {
+			logger.WithError(err).Error("failed to flush output")
+		}
+	case synchronize:
+		for i, commit := range commits {
+			commitLogger := logger.WithField("commit", commit.Hash)
+			commitLogger.Infof("cherry-picking commit %d/%d", i+1, len(commits))
+			if err := cherryPick(ctx, commitLogger, commit); err != nil {
+				logger.WithError(err).Error("failed to cherry-pick commit")
+				break
+			}
+		}
+	}
+}
+
+type commit struct {
+	Date    time.Time `json:"date"`
+	Hash    string    `json:"hash,omitempty"`
+	Author  string    `json:"author,omitempty"`
+	Message string    `json:"message,omitempty"`
+	Repo    string    `json:"repo,omitempty"`
+}
+
+var repoRegex = regexp.MustCompile(`Upstream-repository: ([^ ]+)\n`)
+var commitRegex = regexp.MustCompile(`Upstream-commit: ([a-f0-9]+)\n`)
+
+func detectNewCommits(ctx context.Context, logger *logrus.Entry, stagingDir string) ([]commit, error) {
+	lastCommits := map[string]string{}
+	if err := fs.WalkDir(os.DirFS(stagingDir), ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if d == nil || !d.IsDir() {
+			return nil
+		}
+
+		if path == "." {
+			return nil
+		}
+		logger = logger.WithField("repo", path)
+		logger.Debug("detecting commits")
+		output, err := runCommand(logger, exec.CommandContext(ctx,
+			"git", "log",
+			"-n", "1",
+			"--grep", "Upstream-repository: "+path,
+			"--grep", "Upstream-commit",
+			"--all-match",
+			"--pretty=%B",
+			"--",
+			filepath.Join(stagingDir, path),
```

Comment:
https://github.com/go-git/go-git ?

Reply:
I'm concerned about the `git show --pretty` output and parsing of it.

Reply:
The format used in the sync scripts uses space for separators, but the individual components won't include spaces.

Reply:
What's the specific concern with this command? I pass `"--pretty=%B"` and there's no parsing here.

Reply:
Not this parsing, but the other one... sorry... slightly different view.

Reply:
I amended the other one to use a non-breaking space as the delimiter. I am happy to fix further when automated syncing fails to work after someone uses a non-breaking space in their git commit message.

Reply:
The likelyhood of that is fairly low. I _was_ about to suggest that the author and subject aren't necessary for the processing, but I see you're using them for reporting.